plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/TopQueriesService.java

```diff
@@ -0,0 +1,282 @@
+                    + " should be between ["
+                    + QueryInsightsSettings.MIN_WINDOW_SIZE
+                    + ","
+                    + QueryInsightsSettings.MAX_WINDOW_SIZE
+                    + "]"
+                    + "was ("
+                    + windowSize
+                    + ")"
+            );
+        }
+        if (!(QueryInsightsSettings.VALID_WINDOW_SIZES_IN_MINUTES.contains(windowSize) || windowSize.getMinutes() % 60 == 0)) {
+            throw new IllegalArgumentException(
+                "Window size setting for ["
+                    + metricType
+                    + "]"
+                    + " should be multiple of 1 hour, or one of "
+                    + QueryInsightsSettings.VALID_WINDOW_SIZES_IN_MINUTES
+                    + ", was ("
+                    + windowSize
+                    + ")"
+            );
+        }
+    }
+
+    /**
+     * Get all top queries records that are in the current top n queries store
+     * Optionally include top N records from the last window.
+     *
+     * By default, return the records in sorted order.
+     *
+     * @param includeLastWindow if the top N queries from the last window should be included
+     * @return List of the records that are in the query insight store
+     * @throws IllegalArgumentException if query insight is disabled in the cluster
+     */
+    public List<SearchQueryRecord> getTopQueriesRecords(final boolean includeLastWindow) throws IllegalArgumentException {
+        if (!enabled) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "Cannot get top n queries for [%s] when it is not enabled.", metricType.toString())
+            );
+        }
+        // read from window snapshots
+        final List<SearchQueryRecord> queries = new ArrayList<>(topQueriesCurrentSnapshot.get());
+        if (includeLastWindow) {
+            queries.addAll(topQueriesHistorySnapshot.get());
+        }
+        return Stream.of(queries)
+            .flatMap(Collection::stream)
+            .sorted((a, b) -> SearchQueryRecord.compare(a, b, metricType) * -1)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Consume records to top queries stores
+     *
+     * @param records a list of {@link SearchQueryRecord}
+     */
+    void consumeRecords(final List<SearchQueryRecord> records) {
+        final long currentWindowStart = calculateWindowStart(System.currentTimeMillis());
+        List<SearchQueryRecord> recordsInLastWindow = new ArrayList<>();
+        List<SearchQueryRecord> recordsInThisWindow = new ArrayList<>();
+        for (SearchQueryRecord record : records) {
+            // skip the records that does not have the corresponding measurement
+            if (!record.getMeasurements().containsKey(metricType)) {
+                continue;
+            }
+            if (record.getTimestamp() < currentWindowStart) {
+                recordsInLastWindow.add(record);
```

Comment:
what if its started before the start timestamp of last window? it would be misleading to keep them as part of previous window.

Reply:
We can use "timestamp that the request ends" to decide on the correct window and keep the start time stamp as an attribute in the record. But let me make this as a potential bug and handle it next few days.